snafu/dns_perf_wrapper/trigger_dns_perf.py

```diff
@@ -0,0 +1,117 @@
+from ttp import ttp
+
+logger = logging.getLogger("snafu")
+
+class Trigger_dns_perf():
+    # Initialize the arguments defined
+    def __init__(self, args):
+        self.uuid = args.uuid
+        self.cluster_name = args.cluster_name
+        self.server_address = args.server-address
+        self.queries_per_second = args.queries-per-second
+        self.run_time = args.run-time
+        self.data_file = args.data-file
+        self.clients = args.clients
+
+    # Generate a json payload with the metrics of interest
+    # This is the results document, it gets updated with metrics from the dns-perf run
+    def _json_payload(self, data, timestamp, elapsed_time):
+        #
+        payload = {
+            "uuid": self.uuid,
+            "cluster_name": self.cluster_name,
+            "clients": self.clients,
+            "timestamp": timestamp,
+            "elapsed_time": elapsed_time
+        }
+        try:
+            payload.update(data)
+        except Exception as e:
+            logging.error("Failed to update the result document, error: %s" %(e))
+        return payload
+
+    # Run the dns-perf workload
+    def run_dns_perf_test(self):
+        command = "dnsperf -l {} -s {} -Q {} -d {} -c {}".format(self.run_time, self.server_address, self.queries_per_second, self.data_file, self.clients)
+        logger.info(command)
+        try:
+            out = subprocess.Popen(command, shell=True,
+                                   universal_newlines=True, stdout=subprocess.PIPE)
+            (output, err) = out.communicate()
+            logger.info("Raw result: \n{}".format(output))
+        except Exception as e:
+            logger.error("Failed to run %s, error: %s" % (command, e))
+            exit(1)
+        return output
+
+    # Parse the output generated by dns-perf workload
+    def _parse_stdout(self, stdout):
+       # ttp template to parse the dns-perf stdout
+       template = """
+DNS Performance Testing Tool
+Version {{ dnsperf_version }}
+
+[Status] Command line: {{ command }}
+[Status] Sending queries (to {{ dns_address }})
+[Status] Started at: {{ start_timestamp }}
+[Status] Stopping after {{ stop_time_duration }} seconds
+[Status] Testing complete (time limit)
+
+Statistics:
+
+  Queries sent:         {{ queries_sent }}
+  Queries completed:    {{ queries_completed }} ({{ queries_completed_percentage }})
+  Queries lost:         {{ queries_lost }} ({{ queries_lost_percentage }})
+
+  Response codes:       {{ response_code }}
+  Average packet size:  request {{ avg_request_packet_size }}, response {{ avg_response_packet_size }}
+  Run time (s):         {{ runtime }}
+  Queries per second:   {{ QPS }}
+
+  Average Latency (s):  {{ avg_latency }} (min 0.008983, max 0.017859)
+  Latency StdDev (s):   {{ latency_stddev }}
+        """
+       parser = ttp(stdout, template)
+       parser.parse()
+       parsed_results = parser.result()[0][0]
+       logger.info("Parsed results:")
+       logger.info(parsed_results)
```

Comment:
This is something I am struggling with in a different test. On one-hand, it's useful to dump the data out where it will be read. On the other hand, most standard logs are single-line statements unless an error occurred and you're logging a stack trace. No change needed here -- just something to ponder.